File: applicatie/src/main/java/com/vdzon/administratie/administratie/AdministratieService.java
package com.vdzon.administratie.administratie;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.vdzon.administratie.auth.SessionHelper;
import com.vdzon.administratie.crud.UserCrud;
import com.vdzon.administratie.dto.*;
import com.vdzon.administratie.model.Administratie;
import com.vdzon.administratie.model.AdministratieGegevens;
import com.vdzon.administratie.model.Contact;
import com.vdzon.administratie.model.Gebruiker;
import com.vdzon.administratie.util.SingleAnswer;
import spark.Request;
import spark.Response;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Singleton
public class AdministratieService {

    @Inject
    UserCrud userCrud;

    protected Object putAdministratie(Request req, Response res) throws Exception {
        try {
            String uuid = SessionHelper.getAuthenticatedUserUuid(req);
            if (uuid == null) {
                res.status(404);
                return new SingleAnswer("not found");
            }
            Gebruiker gebruiker = userCrud.getGebruiker(uuid);

            String administratieGegevensJson = req.body();
            ObjectMapper mapper = new ObjectMapper();
            AdministratieGegevensDto administratieGegevensDto = mapper.readValue(administratieGegevensJson, AdministratieGegevensDto.class);
            AdministratieGegevens administratieGegevens = administratieGegevensDto.toAdministratieGegevens();

            Administratie administratie = gebruiker.getDefaultAdministratie().toBuilder().administratieGegevens(administratieGegevens).build();
            ArrayList<Administratie> administraties = new ArrayList<>();
            administraties.add(administratie);
            Gebruiker gebruikerToUpdate = gebruiker.toBuilder().administraties(administraties).build();
            userCrud.updateGebruiker(gebruikerToUpdate);
        } catch (Exception ex) {
            ex.printStackTrace();
            throw ex;
        }
        return new SingleAnswer("ok");
    }
}

File: rest/impl/src/main/java/com/vdzon/administratie/rest/administratie/AdministratieService.kt
package com.vdzon.administratie.rest.administratie

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.vdzon.administratie.authenticatie.AuthenticationService
import com.vdzon.administratie.database.UserDao
import com.vdzon.administratie.dto.AdministratieGegevensDto
import com.vdzon.administratie.util.SingleAnswer
import spark.Request
import spark.Response
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AdministratieService {

    @Inject
    lateinit internal var userDao: UserDao

    @Inject
    lateinit internal var athenticationService: AuthenticationService

    @Throws(Exception::class)
    fun putAdministratie(req: Request, res: Response): Any {
        val gebruiker = athenticationService.getGebruikerOrThowForbiddenException(req, res)

        val administratieGegevensJson = req.body()
        val mapper = jacksonObjectMapper()
        val administratieGegevensDto = mapper.readValue(administratieGegevensJson, AdministratieGegevensDto::class.java)
        val administratieGegevens = administratieGegevensDto.toAdministratieGegevens()

        val administratie = gebruiker.defaultAdministratie.copy(administratieGegevens = administratieGegevens);
        val administraties = Arrays.asList(administratie);
        val gebruikerToUpdate = gebruiker.copy(administraties = administraties)
        userDao!!.updateGebruiker(gebruikerToUpdate);
        return SingleAnswer("ok")
    }

}
